File: Linked_list/lc92_reverse-linked-list-ii.py
"""
92.反转链表 II
给你单链表的头指针 head 和两个整数 left 和 right ，其中 left <= right 。请你反转从位置 left 到位置 right 的链表节点，返回 反转后的链表 。

data:2025-12-25
"""
from typing import Optional
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def reverseBetween(self, head: Optional[ListNode], left: int, right: int) -> Optional[ListNode]:
        dummy = ListNode(0)
        dummy.next = head
        p0 = dummy
        for _ in range(left - 1):
            p0 = p0.next
        pre = None
        cur = p0.next
        for _ in range(right - left + 1):
            nxt = cur.next
            cur.next = pre
            pre = cur
            cur = nxt
        p0.next.next = cur
        p0.next = pre
        return dummy.next
if __name__ == "__main__":
    # 创建示例链表: 1 -> 2 -> 3 -> 4 -> 5
    head = ListNode(1)
    head.next = ListNode(2)
    head.next.next = ListNode(3)
    head.next.next.next = ListNode(4)
    head.next.next.next.next = ListNode(5)

    left, right = 2, 4

    sol = Solution()
    modified_head = sol.reverseBetween(head, left, right)

    # 输出处理后的链表
    current = modified_head
    while current:
        print(current.val, end=" -> " if current.next else "")
        current = current.next
    # 输出: 1 -> 4 -> 3 -> 2 -> 5

File: c_plus_plus/Linked_list/lc92_reverse-linked-list-ii.cpp
/*
92.反转链表2
给你单链表的头指针 head 和两个整数 left 和 right ，其中 left <= right 。请你反转从位置 left 到位置 right 的链表节点，返回 反转后的链表 。

2026.06.08
*/
#include<iostream>
#include<stdio.h>
#include<vector>
using namespace std;


struct ListNode
{
    int val;
    ListNode* next;
    ListNode(int x): val(x),next(nullptr){}
};

ListNode* createList(const vector<int>& nums){
    if(nums.empty()) return nullptr;
    ListNode* head = new ListNode(nums[0]);
    ListNode* cur = head;
    for(int i = 1; i < nums.size(); i++){
        cur->next = new ListNode(nums[i]);
        cur = cur->next;
    }
    return head;
}

void printList(ListNode* head){
    ListNode* cur = head;
    while(cur){
        cout << cur->val;
        if(cur->next!=nullptr) cout << " ";
        cur = cur->next;
    }
    cout << endl;
}

ListNode* reverseBetween(ListNode* head, int left, int right)
{
    ListNode dummy(0);
    dummy.next = head;
    ListNode* p0 = &dummy;
    
    for(int i = 0; i < left - 1; i++){
        p0 = p0->next;
    }
    ListNode* cur = p0->next;
    ListNode* pre = nullptr;
    int len = right - left + 1;
    while(len > 0){
        ListNode* nxt = cur->next;
        cur->next = pre;
        pre = cur;
        cur = nxt;
        len--;
    }
    p0->next->next = cur;
    p0->next = pre;
    return dummy.next;
}

int main()
{
    int n,left,right;
    cin >> n;
    cin >> left >> right;
    vector<int> nums(n);
    for(int i = 0; i < n; i++){
        cin >> nums[i];
    }
    ListNode* head = createList(nums);
    ListNode* dummy = reverseBetween(head, left, right);
    printList(dummy);
    return 0;
}